Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.cpp

```diff
@@ -232,33 +248,45 @@ void LibWebRTCCodecs::setCallbacks(bool useGPUProcess, bool useRemoteFrames)
     webrtc::setVideoEncoderCallbacks(createVideoEncoder, releaseVideoEncoder, initializeVideoEncoder, encodeVideoFrame, registerEncodeCompleteCallback, setEncodeRatesCallback);
 }
 
+// May be called on any thread.
 LibWebRTCCodecs::Decoder* LibWebRTCCodecs::createDecoder(VideoCodecType type)
 {
-    ASSERT(!isMainRunLoop());
+    return createDecoderInternal(type, [](auto&) { });
+}
 
+void LibWebRTCCodecs::createDecoderAndWaitUntilReady(VideoCodecType type, Function<void(Decoder&)>&& callback)
+{
+    createDecoderInternal(type, WTFMove(callback));
+}
+
+LibWebRTCCodecs::Decoder* LibWebRTCCodecs::createDecoderInternal(VideoCodecType type, Function<void(Decoder&)>&& callback)
+{
     auto decoder = makeUnique<Decoder>();
     auto* result = decoder.get();
     decoder->identifier = VideoDecoderIdentifier::generateThreadSafe();
     decoder->type = type;
 
-    ensureGPUProcessConnectionAndDispatchToThread([this, decoder = WTFMove(decoder)]() mutable {
+    ensureGPUProcessConnectionAndDispatchToThread([this, decoder = WTFMove(decoder), callback = WTFMove(callback)]() mutable {
         assertIsCurrent(workQueue());
 
+        auto* decodePointer = decoder.get();
+
         Locker locker { m_connectionLock };
```

Comment:
Do you need to hold the lock while calling the callback?

Reply:
Good point, I'll change this.